### PR TITLE
Add cursor grab for web target

### PR DESCRIPTION
**Pull request by TotalKrill**
- [x] Tested on all platforms changed                                patched winit, tried it in bevy_winit, works
- [x] Compilation warnings were addressed                      cargo build --target wasm32-unknown-unknown
- [x] `cargo fmt` has been run on this branch
- [x] `cargo doc` builds successfully
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

**Comment by cwfitzgerald** [comment]
I have implemented this independent of winit, and enabling pointer grab causes this call to throw a nasty error: https://github.com/rust-windowing/winit/blob/master/src/platform_impl/web/web_sys/canvas/pointer_handler.rs#L82-L84

> scene-viewer.js:347 panicked at 'Failed to set pointer capture: JsValue(InvalidStateError: Failed to execute 'setPointerCapture' on 'Element': InvalidStateError

**Comment by TotalKrill** [comment]
Would not the correct way be to not have an expect in fallible code? 

The other way is if we could accurately predict that the state was actually that the pointer is already locked, that would solve your specific case, but since I do not know which ways that the pointerlock function can fail, its not a very good fix.

I mean, unfortunately in web, elements are shared in an unsafe manner, making it hard to have full control. As you say, you can do JS calls on the element and rust code will not ( i believe ) have a chance to react to it?

**Comment by cwfitzgerald** [comment]
That sounds like a reasonable plan, I'll fork winit tonight and disable the expect to see what happens.

**Comment by cwfitzgerald** [comment]
@TotalKrill I just checked out your branch and disabled the expect there, and everything works as we expect. So with that one change I think this PR is ready to go.

**Comment by maroider** [comment]
> So with that one change I think this PR is ready to go.

\+ resolving the merge conflict for `CHANGELOG.md`

**Comment by TotalKrill** [comment]
Im actually not entirely sure just removing the expect, I rather dislike silent errors

Would a log::error! be okay to add after the set_pointer_capture() call? like this:

``` rust
                let e = canvas.set_pointer_capture(event.pointer_id());
                if !e.is_ok() {
                    log::error("{:?}", e);
                }
```

**Comment by cwfitzgerald** [comment]
The issue with a error is that the error will occur every single time you click the mouse when the cursor is grabbed. That would unnecessarily spam the logs. I don't think there's a situation where this could fail that we really care that it fails.

**Review by cwfitzgerald — APPROVED** [review]

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (madsmtm) [inline src/platform_impl/web/web_sys/canvas/pointer_handler.rs:82]: Nit: Document why we're swallowing the error. Something similar to what @cwfitzgerald wrote is fine:

> ... the error will occur every single time you click the mouse when the cursor is grabbed. [...] there's [not] a situation where this could fail that we really care that it fails.

**Review by madsmtm — APPROVED** [review]

**Comment by TotalKrill** [comment]
I added a comment there and modified the commit message to be a bit more explicit

File: src/window.rs
Comment (maroider) [inline src/window.rs:802]: Are there any quirks of the pointer lock API that warrant the inclusion of this note? To my knowledge, the `## Platform-specific` section is for quirks that are relevant for users of winit, yet I don't see how this is all that useful to know.

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (maroider) [reply src/platform_impl/web/web_sys/canvas/pointer_handler.rs:86]: Are there any other cases where we expect calling `setPointerCapture` to fail?

File: src/window.rs
Comment (TotalKrill) [reply src/window.rs:802]: I can remove that line if that is blocking this PR

File: src/platform_impl/web/web_sys/canvas/pointer_handler.rs
Comment (TotalKrill) [reply src/platform_impl/web/web_sys/canvas/pointer_handler.rs:86]: I do not know to be honest, so if anyone else has any idea, feel free to suggest them

Comment (maroider) [reply src/platform_impl/web/web_sys/canvas/pointer_handler.rs:86]: So, I looked up [the spec](https://w3c.github.io/pointerevents/#setting-pointer-capture), and it seems like the call will fail if there is no [active pointer](https://w3c.github.io/pointerevents/#dfn-active-pointer) (and it will also fail "If the element is not connected", but that's a weird situation to be in, I think).

File: src/window.rs
Comment (maroider) [reply src/window.rs]: In light of the failure conditions for [`setPointerCapture`](https://w3c.github.io/pointerevents/#setting-pointer-capture), I believe the following should be documented:
> 3. If the element is not connected [DOM], throw an "InvalidStateError" DOMException.
> 4. If this method is invoked while the element's node document [DOM] has a locked element ([PointerLock] pointerLockElement), throw an "InvalidStateError" DOMException.
> 5. If the pointer is not in the active buttons state or the element's node document is not the active document of the pointer, then terminate these steps.

This text should probably not be copied verbatim.

**Comment by maroider** [comment]
After coming back to this issue, I've come to the conclusion that the PR is fine as-is, although I'd like to test it myself before approving it.

**Review by maroider — APPROVED** [review]
Seems to work fine when I tested it.

**Comment by ArturKovacs** [comment]
Note that the chagelog entry will end up under the wrong release after merging.

I suggest merging the latest `master` into this branch and then moving your changelog entry to the topmost section.

**Comment by maroider** [comment]
I rebased the branch on top of master to get rid of the merge conflicts. Hope you don't mind :)

I'll merge this in a bit.

**Comment by TotalKrill** [comment]
I definitely don't mind, in fact I am thankful 👍

**Pull request merged**
